### PR TITLE
Check for inherited attrs in initializeAttrs

### PR DESCRIPTION
**Pull request by alexpdraper**
When an element extends another element, none of its inherited `attrs` become defined. This PR updates `initializeAttrs` to traverse the instances prototype chain to gather property names with the `@attr` decorator from parent elements.

~~~typescript
@controller
class KindaCoolElement extends HTMLElement {
  @attr foo = ''
}

// foo will be available on instances of `VeryCoolElement` and 'data-foo'
// will be in `VeryCoolElement`’s `observedAttributes`
@controller
class VeryCoolElement extends KindaCoolElement {}

VeryCoolElement.observedAttributes // ['data-foo']

const veryCool = document.createElement('very-cool')
document.body.appendChild(veryCool)

veryCool.foo === '' // true
veryCool.foo = 'bar'
veryCool.getAttribute('data-foo') // 'bar'
~~~

### CODE REVIEW

File: src/attr.ts [hunk 1]
```diff
@@ -73,6 +73,12 @@ export function initializeAttrs(instance: HTMLElement, names?: Iterable<string>)
   }
 }
 
+function getAttrNames(instance: HTMLElement | typeof HTMLElement): string[] {
+  if (!instance || instance === HTMLElement) return []
+  const proto = Object.getPrototypeOf(instance)
+  return (attrs.get(proto) || []).concat(getAttrNames(proto))
+}
```
Comment (keithamus) [inline src/attr.ts:80]: I'd be interested in seeing this without recursion. I feel a while loop might be simpler to read here.

Comment (alexpdraper) [reply src/attr.ts]: Yup. That’s better for sure.

**Comment by alexpdraper** [comment]
Anything we can do to move this along?

File: src/attr.ts [hunk 2]
```diff
@@ -73,6 +73,18 @@ export function initializeAttrs(instance: HTMLElement, names?: Iterable<string>)
...
+function getAttrNames(instance: HTMLElement): Set<string> {
+  let names: string[] = []
+  let proto = Object.getPrototypeOf(instance)
+
+  while (proto && proto !== HTMLElement) {
+    names = names.concat(attrs.get(proto) || [])
+    proto = Object.getPrototypeOf(proto)
+  }
+
+  return new Set(names)
```
[the remark this answers is not in the document]
Comment (alexpdraper) [reply src/attr.ts]: I think that’s logical. Sometimes loops get nested! I’ve pushed up a change that uses a `Set` from the start and also adds the inherited attrs to the `observedAttributes`.

**Review by koddsson — APPROVED** [review]
This looks good to me ✨

**Comment by alexpdraper** [comment]
> This looks good to me ✨

Thanks so much! Anything I can do to help get this merged?

**Comment by koddsson** [comment]
> > This looks good to me ✨
> 
> Thanks so much! Anything I can do to help get this merged?

I just forgot to merge and publish 😄

**Pull request merged**